src/bin/upgrade/main.rs

```diff
@@ -0,0 +1,108 @@
+//! `cargo upgrade`
+
+#![warn(missing_docs, missing_debug_implementations, missing_copy_implementations, trivial_casts,
+        trivial_numeric_casts, unsafe_code, unstable_features, unused_import_braces,
+        unused_qualifications)]
+
+extern crate docopt;
+extern crate pad;
+#[macro_use]
+extern crate serde_derive;
+extern crate toml;
+
+use std::error::Error;
+use std::io::{self, Write};
+use std::process;
+
+extern crate cargo_edit;
+use cargo_edit::{Manifest, get_latest_dependency};
+
+static USAGE: &'static str = r"
+Upgrade all dependencies in a manifest file to the latest version.
+
+Usage:
+    cargo upgrade [--dependency <dep>...] [--manifest-path <path>]
+    cargo upgrade (-h | --help)
+    cargo upgrade (-V | --version)
+
+Options:
+    -d --dependency <dep>       Specific dependency to upgrade. If this option is used, only the
+                                specified dependencies will be upgraded.
+    --manifest-path <path>      Path to the manifest to upgrade.
+    -h --help                   Show this help page.
+    -V --version                Show version.
+
+Dev, build, and all target dependencies will also be upgraded. Only dependencies from crates.io are
+supported. Git/path dependencies will be ignored.
+";
+
+/// Docopts input args.
+#[derive(Debug, Deserialize)]
+struct Args {
+    /// `--dependency -d <dep>`
+    flag_dependency: Vec<String>,
+    /// `--manifest-path <path>`
+    flag_manifest_path: Option<String>,
+    /// `--version`
+    flag_version: bool,
+}
+
+fn is_version_dependency(dep: &toml::Value) -> bool {
+    dep.as_table()
+        .map(|table| {
+            // Not a version dependency if the `git` or `path` keys are present.
+            !(table.contains_key("git") || table.contains_key("path"))
+        })
+        .unwrap_or(true)
+}
+
+fn update_manifest(
+    manifest_path: &Option<String>,
+    only_update: &[String],
+) -> Result<(), Box<Error>> {
+    let manifest_path = manifest_path.as_ref().map(From::from);
+    let mut manifest = Manifest::open(&manifest_path).unwrap();
+
+    // Look for dependencies in all sections.
+    for (table_path, table) in manifest.get_sections() {
+        table
+            .iter()
+            .filter(|&(name, _old_value)| {
+                // If the user specifies a list of dependencies, only update those dependencies.
+                only_update.is_empty() || only_update.contains(name)
+            })
+            .filter(|&(_name, old_value)| is_version_dependency(old_value))
+            .map(|(name, _old_value)| {
+
+                let latest_version = get_latest_dependency(name, false)?;
+
+                manifest.update_table_entry(&table_path, &latest_version)?;
+
+                Ok(())
+            })
+            .collect::<Result<Vec<_>, Box<Error>>>()?;
```

Comment:
You are collecting this into a Vec… just to drop it? Seems weird. Better use an inner for-loop with the body of the map (this way the `?` will return in the fn scope)